File: apps/android/core/src/main/kotlin/com/discountscreener/core/engine/ReportingEngine.kt
package com.discountscreener.core.engine

import com.discountscreener.core.model.AlertEvent
import com.discountscreener.core.model.AlertKind
import com.discountscreener.core.model.CandidateRow
import com.discountscreener.core.model.ConfidenceBand
import com.discountscreener.core.model.ExternalSignalStatus
import com.discountscreener.core.model.ExternalValuationSignal
import com.discountscreener.core.model.FundamentalSnapshot
import com.discountscreener.core.model.MarketSnapshot
import com.discountscreener.core.model.PersistedReportState
import com.discountscreener.core.model.PersistedSymbolState
import com.discountscreener.core.model.PriceHistoryPoint
import com.discountscreener.core.model.QualificationStatus
import com.discountscreener.core.model.SymbolDetail
import com.discountscreener.core.model.TapeEvent
import com.discountscreener.core.model.ViewFilter
import java.math.BigInteger

private const val MAX_PRICE_HISTORY_PER_SYMBOL = 240

class ReportingEngine(
    private val minGapBps: Int = 2_000,
    private val externalSignalMaxAgeSeconds: Long = 30,
    private val tapeCapacity: Int = 8,
) {
    private data class SymbolState(
        var snapshot: MarketSnapshot? = null,
        var externalSignal: ExternalValuationSignal? = null,
        var fundamentals: FundamentalSnapshot? = null,
        var lastSequence: Int = 0,
        var updateCount: Int = 0,
        val priceHistory: ArrayDeque<PriceHistoryPoint> = ArrayDeque(),
    )

    private val symbols = linkedMapOf<String, SymbolState>()
    private val watchlist = linkedSetOf<String>()
    private val recentTape = ArrayDeque<TapeEvent>()
    private val recentAlerts = ArrayDeque<AlertEvent>()

    var totalEvents: Int = 0
        private set

    var latestSequence: Int = 0
        private set

    fun symbolCount(): Int = symbols.size

    fun trackedSymbols(): List<String> = symbols.keys.toList()

    fun ingestSnapshot(snapshot: MarketSnapshot) {
        if (snapshot.marketPriceCents <= 0 || snapshot.intrinsicValueCents <= 0) {
            return
        }

        val previousDetail = detail(snapshot.symbol)
        val sequence = nextSequence()
        val state = symbols.getOrPut(snapshot.symbol) { SymbolState() }
        state.snapshot = snapshot
        state.lastSequence = sequence
        state.updateCount += 1
        while (state.priceHistory.size >= MAX_PRICE_HISTORY_PER_SYMBOL) {
            state.priceHistory.removeFirst()
        }
        state.priceHistory.addLast(
            PriceHistoryPoint(sequence = sequence, marketPriceCents = snapshot.marketPriceCents),
        )
        totalEvents += 1
        pushTape(snapshot.symbol)
        pushAlerts(snapshot.symbol, previousDetail)
    }

    fun ingestExternal(signal: ExternalValuationSignal) {
        if (signal.fairValueCents <= 0) {
            return
        }

        val sanitized = sanitizeExternalSignal(signal)
        val previousDetail = detail(sanitized.symbol)
        val sequence = nextSequence()
        val state = symbols.getOrPut(sanitized.symbol) { SymbolState() }
        state.externalSignal = sanitized
        state.lastSequence = sequence
        state.updateCount += 1
        totalEvents += 1
        pushTape(sanitized.symbol)
        pushAlerts(sanitized.symbol, previousDetail)
    }

    fun ingestFundamentals(fundamentals: FundamentalSnapshot) {
        if (!fundamentals.hasAnyValues()) {
            return
        }

        val sequence = nextSequence()
        val state = symbols.getOrPut(fundamentals.symbol) { SymbolState() }
        state.fundamentals = fundamentals
        state.lastSequence = sequence
        state.updateCount += 1
        totalEvents += 1
    }

    fun candidate(symbol: String): CandidateRow? = symbols[symbol]?.let(::buildCandidate)

    fun detail(symbol: String): SymbolDetail? = symbols[symbol]?.let(::buildDetail)

    fun topRows(limit: Int): List<CandidateRow> = sortedRows().take(limit)

    fun filteredRows(limit: Int = Int.MAX_VALUE, filter: ViewFilter = ViewFilter()): List<CandidateRow> {
        val query = filter.query.trim()
        val rows = symbols
            .filter { (symbol, _) ->
                val queryMatches = query.isEmpty() || symbol.contains(query, ignoreCase = true)
                val watchlistMatches = !filter.watchlistOnly || watchlist.contains(symbol)
                queryMatches && watchlistMatches
            }
            .values
            .mapNotNull(::buildCandidate)
            .toMutableList()
        sortRows(rows)
        return rows.take(limit)
    }

    fun priceHistory(symbol: String, limit: Int = MAX_PRICE_HISTORY_PER_SYMBOL): List<PriceHistoryPoint> =
        symbols[symbol]
            ?.priceHistory
            ?.toList()
            ?.takeLast(limit)
            ?: emptyList()

    fun watchlistSymbols(): List<String> = watchlist.toList().sorted()

    fun replaceWatchlist(symbols: List<String>) {
        watchlist.clear()
        watchlist.addAll(symbols)
    }

    fun toggleWatchlist(symbol: String): Boolean {
        return if (watchlist.remove(symbol)) {
            false
        } else {
            watchlist.add(symbol)
            true
        }
    }

    fun isWatched(symbol: String): Boolean = watchlist.contains(symbol)

    fun recentTape(): List<TapeEvent> = recentTape.toList()

    fun alerts(): List<AlertEvent> = recentAlerts.toList()

    fun restore(state: PersistedReportState) {
        totalEvents = 0
        latestSequence = 0
        symbols.clear()
        recentTape.clear()
        recentAlerts.clear()
        watchlist.clear()

        state.symbolStates.forEach { persisted ->
            latestSequence = maxOf(latestSequence, persisted.lastSequence)
            totalEvents += persisted.updateCount
            val priceHistory = ArrayDeque<PriceHistoryPoint>()
            persisted.priceHistory.take(MAX_PRICE_HISTORY_PER_SYMBOL).forEach(priceHistory::addLast)
            symbols[persisted.symbol] = SymbolState(
                snapshot = persisted.snapshot,
                externalSignal = persisted.externalSignal?.let(::sanitizeExternalSignal),
                fundamentals = persisted.fundamentals,
                lastSequence = persisted.lastSequence,
                updateCount = persisted.updateCount,
                priceHistory = priceHistory,
            )
        }
        watchlist.addAll(state.watchlist)
    }

    fun persistedState(): List<PersistedSymbolState> = symbols
        .map { (symbol, state) ->
            PersistedSymbolState(
                symbol = symbol,
                snapshot = state.snapshot,
                externalSignal = state.externalSignal,
                fundamentals = state.fundamentals,
                lastSequence = state.lastSequence,
                updateCount = state.updateCount,
                priceHistory = state.priceHistory.toList(),
            )
        }
        .sortedBy { it.symbol }

    private fun buildCandidate(state: SymbolState): CandidateRow? {
        val detail = buildDetail(state) ?: return null
        return CandidateRow(
            symbol = detail.symbol,
            marketPriceCents = detail.marketPriceCents,
            intrinsicValueCents = detail.intrinsicValueCents,
            gapBps = detail.gapBps,
            upsideBps = detail.upsideBps,
            isQualified = detail.qualification == QualificationStatus.Qualified,
            confidence = detail.confidence,
            companyName = detail.companyName,
        )
    }

    private fun buildDetail(state: SymbolState): SymbolDetail? =
        buildSymbolDetail(
            snapshot = state.snapshot,
            externalSignal = state.externalSignal,
            fundamentals = state.fundamentals,
            minGapBps = minGapBps,
            externalSignalMaxAgeSeconds = externalSignalMaxAgeSeconds,
            lastSequence = state.lastSequence,
            updateCount = state.updateCount,
            isWatched = watchlist.contains(state.snapshot?.symbol),
        )

    private fun sortedRows(): List<CandidateRow> = buildList {
        symbols.values.mapNotNullTo(this, ::buildCandidate)
    }.toMutableList().also(::sortRows)

    private fun sortRows(rows: MutableList<CandidateRow>) {
        rows.sortWith(
            compareByDescending<CandidateRow> { it.isQualified }
                .thenByDescending { it.upsideBps }
                .thenByDescending { confidenceRank(it.confidence) }
                .thenBy { it.symbol },
        )
    }

    private fun confidenceRank(confidence: ConfidenceBand): Int = when (confidence) {
        ConfidenceBand.Low -> 0
        ConfidenceBand.Provisional -> 1
        ConfidenceBand.High -> 2
    }

    private fun nextSequence(): Int {
        latestSequence += 1
        return latestSequence
    }

    private fun pushTape(symbol: String) {
        if (tapeCapacity <= 0) return
        val candidate = candidate(symbol) ?: return
        while (recentTape.size >= tapeCapacity) {
            recentTape.removeFirst()
        }
        recentTape.addLast(
            TapeEvent(
                symbol = candidate.symbol,
                gapBps = candidate.gapBps,
                isQualified = candidate.isQualified,
                confidence = candidate.confidence,
            ),
        )
    }

    private fun pushAlerts(symbol: String, previousDetail: SymbolDetail?) {
        val currentDetail = detail(symbol) ?: return
        val previousQualified = previousDetail?.qualification == QualificationStatus.Qualified
        val currentQualified = currentDetail.qualification == QualificationStatus.Qualified

        when {
            !previousQualified && currentQualified -> pushAlert(symbol, AlertKind.EnteredQualified, currentDetail.lastSequence)
            previousQualified && !currentQualified -> pushAlert(symbol, AlertKind.ExitedQualified, currentDetail.lastSequence)
            previousQualified && currentQualified &&
                previousDetail?.confidence != ConfidenceBand.High &&
                currentDetail.confidence == ConfidenceBand.High ->
                pushAlert(symbol, AlertKind.ConfidenceUpgraded, currentDetail.lastSequence)
        }
    }

    private fun pushAlert(symbol: String, kind: AlertKind, sequence: Int) {
        if (tapeCapacity <= 0) return
        while (recentAlerts.size >= tapeCapacity) {
            recentAlerts.removeFirst()
        }
        recentAlerts.addLast(AlertEvent(symbol = symbol, kind = kind, sequence = sequence))
    }
}

fun checkedGapBps(marketPriceCents: Long, fairValueCents: Long): Int? {
    if (fairValueCents <= 0) {
        return null
    }
    val fair = BigInteger.valueOf(fairValueCents)
    val market = BigInteger.valueOf(marketPriceCents)
    val scaledGapBps = ((fair - market) * BigInteger.valueOf(10_000L)) / fair
    return scaledGapBps.coerceIn(BigInteger.valueOf(Int.MIN_VALUE.toLong()), BigInteger.valueOf(Int.MAX_VALUE.toLong())).toInt()
}

fun buildSymbolDetail(
    snapshot: MarketSnapshot?,
    externalSignal: ExternalValuationSignal?,
    fundamentals: FundamentalSnapshot?,
    minGapBps: Int = 2_000,
    externalSignalMaxAgeSeconds: Long = 30,
    lastSequence: Int = 0,
    updateCount: Int = 0,
    isWatched: Boolean = false,
): SymbolDetail? {
    snapshot ?: return null
    val sanitizedExternal = externalSignal?.let(::sanitizeExternalSignal)
    val internalGapBps = checkedGapBps(snapshot.marketPriceCents, snapshot.intrinsicValueCents) ?: 0
    val internalUpsideBps = checkedUpsideBps(snapshot.marketPriceCents, snapshot.intrinsicValueCents) ?: 0
    val qualification = qualificationFor(snapshot, internalGapBps, minGapBps)
    val externalStatus = externalStatusFor(snapshot, sanitizedExternal, minGapBps, externalSignalMaxAgeSeconds)
    val confidence = confidenceFor(qualification, externalStatus)
    val weightedFairValue = clampedWeightedFairValue(sanitizedExternal)

    return SymbolDetail(
        symbol = snapshot.symbol,
        profitable = snapshot.profitable,
        marketPriceCents = snapshot.marketPriceCents,
        intrinsicValueCents = snapshot.intrinsicValueCents,
        gapBps = internalGapBps,
        upsideBps = internalUpsideBps,
        minimumGapBps = minGapBps,
        qualification = qualification,
        externalStatus = externalStatus,
        externalSignalFairValueCents = sanitizedExternal?.fairValueCents,
        externalSignalLowFairValueCents = sanitizedExternal?.lowFairValueCents,
        externalSignalHighFairValueCents = sanitizedExternal?.highFairValueCents,
        weightedExternalSignalFairValueCents = weightedFairValue,
        weightedAnalystCount = if (weightedFairValue != null) sanitizedExternal?.weightedAnalystCount else null,
        externalSignalGapBps = sanitizedExternal?.let { checkedGapBps(snapshot.marketPriceCents, it.fairValueCents) },
        externalSignalAgeSeconds = sanitizedExternal?.ageSeconds,
        externalSignalMaxAgeSeconds = externalSignalMaxAgeSeconds,
        analystOpinionCount = sanitizedExternal?.analystOpinionCount,
        recommendationMeanHundredths = sanitizedExternal?.recommendationMeanHundredths,
        strongBuyCount = sanitizedExternal?.strongBuyCount,
        buyCount = sanitizedExternal?.buyCount,
        holdCount = sanitizedExternal?.holdCount,
        sellCount = sanitizedExternal?.sellCount,
        strongSellCount = sanitizedExternal?.strongSellCount,
        fundamentals = fundamentals,
        confidence = confidence,
        lastSequence = lastSequence,
        updateCount = updateCount,
        isWatched = isWatched,
        companyName = snapshot.companyName,
    )
}

private fun qualificationFor(
    snapshot: MarketSnapshot,
    gapBps: Int,
    minGapBps: Int,
): QualificationStatus = when {
    !snapshot.profitable -> QualificationStatus.Unprofitable
    gapBps >= minGapBps -> QualificationStatus.Qualified
    else -> QualificationStatus.GapTooSmall
}

private fun externalStatusFor(
    snapshot: MarketSnapshot,
    externalSignal: ExternalValuationSignal?,
    minGapBps: Int,
    externalSignalMaxAgeSeconds: Long,
): ExternalSignalStatus {
    externalSignal ?: return ExternalSignalStatus.Missing
    if (externalSignal.symbol != snapshot.symbol) return ExternalSignalStatus.Divergent
    if (externalSignal.ageSeconds > externalSignalMaxAgeSeconds) return ExternalSignalStatus.Stale
    return if ((checkedGapBps(snapshot.marketPriceCents, externalSignal.fairValueCents) ?: Int.MIN_VALUE) >= minGapBps) {
        ExternalSignalStatus.Supportive
    } else {
        ExternalSignalStatus.Divergent
    }
}

private fun confidenceFor(
    qualification: QualificationStatus,
    externalStatus: ExternalSignalStatus,
): ConfidenceBand {
    if (qualification != QualificationStatus.Qualified) {
        return ConfidenceBand.Low
    }
    return when (externalStatus) {
        ExternalSignalStatus.Missing -> ConfidenceBand.Provisional
        ExternalSignalStatus.Supportive -> ConfidenceBand.High
        ExternalSignalStatus.Stale, ExternalSignalStatus.Divergent -> ConfidenceBand.Low
    }
}

fun checkedUpsideBps(marketPriceCents: Long, fairValueCents: Long): Int? {
    if (marketPriceCents <= 0 || fairValueCents <= 0) {
        return null
    }
    val fair = BigInteger.valueOf(fairValueCents)
    val market = BigInteger.valueOf(marketPriceCents)
    val scaledUpsideBps = ((fair - market) * BigInteger.valueOf(10_000L)) / market
    return scaledUpsideBps.coerceIn(BigInteger.valueOf(Int.MIN_VALUE.toLong()), BigInteger.valueOf(Int.MAX_VALUE.toLong())).toInt()
}

fun clampedWeightedFairValue(signal: ExternalValuationSignal?): Long? {
    signal ?: return null
    var weighted = signal.weightedFairValueCents ?: return null
    if (signal.lowFairValueCents != null && signal.highFairValueCents != null) {
        val low = minOf(signal.lowFairValueCents, signal.highFairValueCents)
        val high = maxOf(signal.lowFairValueCents, signal.highFairValueCents)
        weighted = weighted.coerceIn(low, high)
    }
    return weighted.takeIf { it > 0 }
}

fun sanitizeExternalSignal(signal: ExternalValuationSignal): ExternalValuationSignal {
    val weighted = clampedWeightedFairValue(signal)
    return if (weighted == null) {
        signal.copy(weightedFairValueCents = null, weightedAnalystCount = null)
    } else {
        signal.copy(weightedFairValueCents = weighted)
    }
}

File: apps/flutter/packages/ds_core/lib/src/reporting_engine.dart
import 'models.dart';

const int maxPriceHistoryPerSymbol = 240;

class ReportingEngine {
  ReportingEngine({
    this.minGapBps = 2000,
    this.externalSignalMaxAgeSeconds = 30,
    this.tapeCapacity = 8,
  });

  final int minGapBps;
  final int externalSignalMaxAgeSeconds;
  final int tapeCapacity;

  final Map<String, _SymbolState> _symbols = {};
  final Set<String> _watchlist = {};
  final List<TapeEvent> _recentTape = [];
  final List<AlertEvent> _recentAlerts = [];

  int totalEvents = 0;
  int latestSequence = 0;

  int symbolCount() => _symbols.length;

  List<String> trackedSymbols() => _symbols.keys.toList();

  void ingestSnapshot(MarketSnapshot snapshot) {
    if (snapshot.marketPriceCents <= 0 || snapshot.intrinsicValueCents <= 0) {
      return;
    }
    final previousDetail = detail(snapshot.symbol);
    final sequence = _nextSequence();
    final state = _symbols.putIfAbsent(snapshot.symbol, _SymbolState.new);
    state.snapshot = snapshot;
    state.lastSequence = sequence;
    state.updateCount += 1;
    while (state.priceHistory.length >= maxPriceHistoryPerSymbol) {
      state.priceHistory.removeAt(0);
    }
    state.priceHistory.add(
      PriceHistoryPoint(sequence: sequence, marketPriceCents: snapshot.marketPriceCents),
    );
    totalEvents += 1;
    _pushTape(snapshot.symbol);
    _pushAlerts(snapshot.symbol, previousDetail);
  }

  void ingestExternal(ExternalValuationSignal signal) {
    if (signal.fairValueCents <= 0) return;
    final sanitized = sanitizeExternalSignal(signal);
    final previousDetail = detail(sanitized.symbol);
    final sequence = _nextSequence();
    final state = _symbols.putIfAbsent(sanitized.symbol, _SymbolState.new);
    state.externalSignal = sanitized;
    state.lastSequence = sequence;
    state.updateCount += 1;
    totalEvents += 1;
    _pushTape(sanitized.symbol);
    _pushAlerts(sanitized.symbol, previousDetail);
  }

  void ingestFundamentals(FundamentalSnapshot fundamentals) {
    if (!fundamentals.hasAnyValues()) return;
    final sequence = _nextSequence();
    final state = _symbols.putIfAbsent(fundamentals.symbol, _SymbolState.new);
    state.fundamentals = fundamentals;
    state.lastSequence = sequence;
    state.updateCount += 1;
    totalEvents += 1;
  }

  CandidateRow? candidate(String symbol) {
    final state = _symbols[symbol];
    return state == null ? null : _buildCandidate(state);
  }

  SymbolDetail? detail(String symbol) {
    final state = _symbols[symbol];
    return state == null ? null : _buildDetail(state);
  }

  List<CandidateRow> topRows(int limit) => _sortedRows().take(limit).toList();

  List<CandidateRow> filteredRows({
    int limit = 0x7fffffff,
    ViewFilter filter = const ViewFilter(),
  }) {
    final query = filter.query.trim();
    final rows = <CandidateRow>[];
    for (final entry in _symbols.entries) {
      final symbol = entry.key;
      final queryMatches =
          query.isEmpty || symbol.toUpperCase().contains(query.toUpperCase());
      final watchlistMatches = !filter.watchlistOnly || _watchlist.contains(symbol);
      if (!queryMatches || !watchlistMatches) continue;
      final row = _buildCandidate(entry.value);
      if (row != null) rows.add(row);
    }
    _sortRows(rows);
    if (rows.length > limit) return rows.sublist(0, limit);
    return rows;
  }

  List<PriceHistoryPoint> priceHistory(String symbol, {int limit = maxPriceHistoryPerSymbol}) {
    final history = _symbols[symbol]?.priceHistory ?? const [];
    if (history.length <= limit) return List.of(history);
    return history.sublist(history.length - limit);
  }

  List<String> watchlistSymbols() => (_watchlist.toList()..sort());

  void replaceWatchlist(List<String> symbols) {
    _watchlist
      ..clear()
      ..addAll(symbols);
  }

  bool toggleWatchlist(String symbol) {
    if (_watchlist.remove(symbol)) return false;
    _watchlist.add(symbol);
    return true;
  }

  bool isWatched(String symbol) => _watchlist.contains(symbol);

  List<TapeEvent> recentTape() => List.of(_recentTape);

  List<AlertEvent> alerts() => List.of(_recentAlerts);

  void restore(PersistedReportState state) {
    totalEvents = 0;
    latestSequence = 0;
    _symbols.clear();
    _recentTape.clear();
    _recentAlerts.clear();
    _watchlist.clear();

    for (final persisted in state.symbolStates) {
      latestSequence = latestSequence > persisted.lastSequence
          ? latestSequence
          : persisted.lastSequence;
      totalEvents += persisted.updateCount;
      final history = persisted.priceHistory.take(maxPriceHistoryPerSymbol).toList();
      _symbols[persisted.symbol] = _SymbolState()
        ..snapshot = persisted.snapshot
        ..externalSignal =
            persisted.externalSignal == null ? null : sanitizeExternalSignal(persisted.externalSignal!)
        ..fundamentals = persisted.fundamentals
        ..lastSequence = persisted.lastSequence
        ..updateCount = persisted.updateCount
        ..priceHistory.addAll(history);
    }
    _watchlist.addAll(state.watchlist);
  }

  List<PersistedSymbolState> persistedState() {
    final list = _symbols.entries
        .map(
          (e) => PersistedSymbolState(
            symbol: e.key,
            snapshot: e.value.snapshot,
            externalSignal: e.value.externalSignal,
            fundamentals: e.value.fundamentals,
            lastSequence: e.value.lastSequence,
            updateCount: e.value.updateCount,
            priceHistory: List.of(e.value.priceHistory),
          ),
        )
        .toList()
      ..sort((a, b) => a.symbol.compareTo(b.symbol));
    return list;
  }

  CandidateRow? _buildCandidate(_SymbolState state) {
    final d = _buildDetail(state);
    if (d == null) return null;
    return CandidateRow(
      symbol: d.symbol,
      marketPriceCents: d.marketPriceCents,
      intrinsicValueCents: d.intrinsicValueCents,
      gapBps: d.gapBps,
      upsideBps: d.upsideBps,
      isQualified: d.qualification == QualificationStatus.qualified,
      confidence: d.confidence,
      companyName: d.companyName,
    );
  }

  SymbolDetail? _buildDetail(_SymbolState state) => buildSymbolDetail(
        snapshot: state.snapshot,
        externalSignal: state.externalSignal,
        fundamentals: state.fundamentals,
        minGapBps: minGapBps,
        externalSignalMaxAgeSeconds: externalSignalMaxAgeSeconds,
        lastSequence: state.lastSequence,
        updateCount: state.updateCount,
        isWatched: state.snapshot != null && _watchlist.contains(state.snapshot!.symbol),
      );

  List<CandidateRow> _sortedRows() {
    final rows = <CandidateRow>[];
    for (final state in _symbols.values) {
      final row = _buildCandidate(state);
      if (row != null) rows.add(row);
    }
    _sortRows(rows);
    return rows;
  }

  void _sortRows(List<CandidateRow> rows) {
    rows.sort((a, b) {
      final q = (b.isQualified ? 1 : 0).compareTo(a.isQualified ? 1 : 0);
      if (q != 0) return q;
      final u = b.upsideBps.compareTo(a.upsideBps);
      if (u != 0) return u;
      final c = _confidenceRank(b.confidence).compareTo(_confidenceRank(a.confidence));
      if (c != 0) return c;
      return a.symbol.compareTo(b.symbol);
    });
  }

  int _confidenceRank(ConfidenceBand confidence) => switch (confidence) {
        ConfidenceBand.low => 0,
        ConfidenceBand.provisional => 1,
        ConfidenceBand.high => 2,
      };

  int _nextSequence() {
    latestSequence += 1;
    return latestSequence;
  }

  void _pushTape(String symbol) {
    if (tapeCapacity <= 0) return;
    final c = candidate(symbol);
    if (c == null) return;
    while (_recentTape.length >= tapeCapacity) {
      _recentTape.removeAt(0);
    }
    _recentTape.add(
      TapeEvent(
        symbol: c.symbol,
        gapBps: c.gapBps,
        isQualified: c.isQualified,
        confidence: c.confidence,
      ),
    );
  }

  void _pushAlerts(String symbol, SymbolDetail? previousDetail) {
    final currentDetail = detail(symbol);
    if (currentDetail == null) return;
    final previousQualified = previousDetail?.qualification == QualificationStatus.qualified;
    final currentQualified = currentDetail.qualification == QualificationStatus.qualified;

    if (!previousQualified && currentQualified) {
      _pushAlert(symbol, AlertKind.enteredQualified, currentDetail.lastSequence);
    } else if (previousQualified && !currentQualified) {
      _pushAlert(symbol, AlertKind.exitedQualified, currentDetail.lastSequence);
    } else if (previousQualified &&
        currentQualified &&
        previousDetail?.confidence != ConfidenceBand.high &&
        currentDetail.confidence == ConfidenceBand.high) {
      _pushAlert(symbol, AlertKind.confidenceUpgraded, currentDetail.lastSequence);
    }
  }

  void _pushAlert(String symbol, AlertKind kind, int sequence) {
    if (tapeCapacity <= 0) return;
    while (_recentAlerts.length >= tapeCapacity) {
      _recentAlerts.removeAt(0);
    }
    _recentAlerts.add(AlertEvent(symbol: symbol, kind: kind, sequence: sequence));
  }
}

class _SymbolState {
  MarketSnapshot? snapshot;
  ExternalValuationSignal? externalSignal;
  FundamentalSnapshot? fundamentals;
  int lastSequence = 0;
  int updateCount = 0;
  final List<PriceHistoryPoint> priceHistory = [];
}

int? checkedGapBps(int marketPriceCents, int fairValueCents) {
  if (fairValueCents <= 0) return null;
  final scaled = ((BigInt.from(fairValueCents) - BigInt.from(marketPriceCents)) *
          BigInt.from(10000)) ~/
      BigInt.from(fairValueCents);
  return scaled.toInt().clamp(-0x80000000, 0x7fffffff);
}

int? checkedUpsideBps(int marketPriceCents, int fairValueCents) {
  if (marketPriceCents <= 0 || fairValueCents <= 0) return null;
  final scaled = ((BigInt.from(fairValueCents) - BigInt.from(marketPriceCents)) *
          BigInt.from(10000)) ~/
      BigInt.from(marketPriceCents);
  return scaled.toInt().clamp(-0x80000000, 0x7fffffff);
}

int? clampedWeightedFairValue(ExternalValuationSignal? signal) {
  if (signal == null) return null;
  var weighted = signal.weightedFairValueCents;
  if (weighted == null) return null;
  if (signal.lowFairValueCents != null && signal.highFairValueCents != null) {
    final low = signal.lowFairValueCents! < signal.highFairValueCents!
        ? signal.lowFairValueCents!
        : signal.highFairValueCents!;
    final high = signal.lowFairValueCents! > signal.highFairValueCents!
        ? signal.lowFairValueCents!
        : signal.highFairValueCents!;
    weighted = weighted.clamp(low, high);
  }
  return weighted > 0 ? weighted : null;
}

ExternalValuationSignal sanitizeExternalSignal(ExternalValuationSignal signal) {
  final weighted = clampedWeightedFairValue(signal);
  if (weighted == null) {
    return signal.copyWith(clearWeighted: true);
  }
  return signal.copyWith(weightedFairValueCents: weighted);
}

SymbolDetail? buildSymbolDetail({
  MarketSnapshot? snapshot,
  ExternalValuationSignal? externalSignal,
  FundamentalSnapshot? fundamentals,
  int minGapBps = 2000,
  int externalSignalMaxAgeSeconds = 30,
  int lastSequence = 0,
  int updateCount = 0,
  bool isWatched = false,
}) {
  if (snapshot == null) return null;
  final sanitizedExternal =
      externalSignal == null ? null : sanitizeExternalSignal(externalSignal);
  // Model intrinsic drives gap/upside/qualification (shared contract). UI
  // fair value is reprojected at ingest via FairValueSelection (analyst-first).
  final internalGapBps =
      checkedGapBps(snapshot.marketPriceCents, snapshot.intrinsicValueCents) ?? 0;
  final internalUpsideBps =
      checkedUpsideBps(snapshot.marketPriceCents, snapshot.intrinsicValueCents) ?? 0;
  final qualification = _qualificationFor(snapshot, internalGapBps, minGapBps);
  final externalStatus = _externalStatusFor(
    snapshot,
    sanitizedExternal,
    minGapBps,
    externalSignalMaxAgeSeconds,
  );
  final confidence = _confidenceFor(qualification, externalStatus, sanitizedExternal);
  final weightedFairValue = clampedWeightedFairValue(sanitizedExternal);

  return SymbolDetail(
    symbol: snapshot.symbol,
    profitable: snapshot.profitable,
    marketPriceCents: snapshot.marketPriceCents,
    intrinsicValueCents: snapshot.intrinsicValueCents,
    gapBps: internalGapBps,
    upsideBps: internalUpsideBps,
    minimumGapBps: minGapBps,
    qualification: qualification,
    externalStatus: externalStatus,
    externalSignalFairValueCents: sanitizedExternal?.fairValueCents,
    externalSignalLowFairValueCents: sanitizedExternal?.lowFairValueCents,
    externalSignalHighFairValueCents: sanitizedExternal?.highFairValueCents,
    weightedExternalSignalFairValueCents: weightedFairValue,
    weightedAnalystCount: weightedFairValue != null ? sanitizedExternal?.weightedAnalystCount : null,
    externalSignalGapBps: sanitizedExternal == null
        ? null
        : checkedGapBps(snapshot.marketPriceCents, sanitizedExternal.fairValueCents),
    externalSignalAgeSeconds: sanitizedExternal?.ageSeconds,
    externalSignalMaxAgeSeconds: externalSignalMaxAgeSeconds,
    analystOpinionCount: sanitizedExternal?.analystOpinionCount,
    recommendationMeanHundredths: sanitizedExternal?.recommendationMeanHundredths,
    strongBuyCount: sanitizedExternal?.strongBuyCount,
    buyCount: sanitizedExternal?.buyCount,
    holdCount: sanitizedExternal?.holdCount,
    sellCount: sanitizedExternal?.sellCount,
    strongSellCount: sanitizedExternal?.strongSellCount,
    fundamentals: fundamentals,
    confidence: confidence,
    lastSequence: lastSequence,
    updateCount: updateCount,
    isWatched: isWatched,
    companyName: snapshot.companyName,
  );
}

QualificationStatus _qualificationFor(
  MarketSnapshot snapshot,
  int gapBps,
  int minGapBps,
) {
  if (!snapshot.profitable) return QualificationStatus.unprofitable;
  if (gapBps >= minGapBps) return QualificationStatus.qualified;
  return QualificationStatus.gapTooSmall;
}

ExternalSignalStatus _externalStatusFor(
  MarketSnapshot snapshot,
  ExternalValuationSignal? externalSignal,
  int minGapBps,
  int externalSignalMaxAgeSeconds,
) {
  if (externalSignal == null) return ExternalSignalStatus.missing;
  if (externalSignal.symbol != snapshot.symbol) return ExternalSignalStatus.divergent;
  if (externalSignal.ageSeconds > externalSignalMaxAgeSeconds) {
    return ExternalSignalStatus.stale;
  }
  final gap =
      checkedGapBps(snapshot.marketPriceCents, externalSignal.fairValueCents) ?? -0x80000000;
  return gap >= minGapBps ? ExternalSignalStatus.supportive : ExternalSignalStatus.divergent;
}

ConfidenceBand _confidenceFor(
  QualificationStatus qualification,
  ExternalSignalStatus externalStatus,
  ExternalValuationSignal? externalSignal,
) {
  if (qualification != QualificationStatus.qualified) return ConfidenceBand.low;
  switch (externalStatus) {
    case ExternalSignalStatus.missing:
      return ConfidenceBand.provisional;
    case ExternalSignalStatus.supportive:
      return _hasHighConfidenceAnalystCoverage(externalSignal)
          ? ConfidenceBand.high
          : ConfidenceBand.provisional;
    case ExternalSignalStatus.stale:
    case ExternalSignalStatus.divergent:
      return ConfidenceBand.low;
  }
}

bool _hasHighConfidenceAnalystCoverage(ExternalValuationSignal? signal) {
  final count = signal?.weightedAnalystCount ?? signal?.analystOpinionCount;
  return count != null && count >= 3;
}
